### PR TITLE
[SECRES-2569] [SECRES-2584] Add a subcommand to configure the firewall environment

### PR DESCRIPTION
**Pull request by ikretz**
This PR adds subcommands to the supply-chain firewall:

```bash
$ scfw -h
usage: scfw [-h] [-v] [--log-level LEVEL] {run,configure} ...

A tool for preventing the installation of malicious PyPI and npm packages.

positional arguments:
  {run,configure}

options:
  -h, --help         show this help message and exit
  -v, --version      show program's version number and exit
  --log-level LEVEL  Desired logging level (default: WARNING, options: DEBUG, INFO, WARNING, ERROR)
```

- The `run` subcommand contains all of the previous firewall functionality: `scfw pip install requests` becomes `scfw run pip install requests`, etc.

```bash
$ scfw run -h
usage: scfw run [options] COMMAND

Run a package manager command through the supply-chain firewall.

options:
  -h, --help         show this help message and exit
  --dry-run          Verify any installation targets but do not run the package manager command
  --executable PATH  Python or npm executable to use for running commands (default: environmentally determined)
```

- The `configure` subcommand allows the user to configure their environment with shell aliases and environment variables for Datadog logging.

```bash
$ scfw configure -h
usage: scfw configure [-h]

Configure the environment for using the supply-chain firewall.

options:
  -h, --help  show this help message and exit
```

It also adds single-source versioning and a `--version` command-line option.

### CODE REVIEW

File: scfw/cli.py [hunk 1]
```diff
@@ -90,8 +132,16 @@ def _parse_command_line(argv: list[str]) -> tuple[Optional[Namespace], str]:
 
     try:
         args = parser.parse_args(argv[1:hinge])
-        args_dict = vars(args)
-        args_dict["command"] = argv[hinge:]
+
+        match args.subcommand == "run", hinge == len(argv):
```
Comment (sobregosodd) [inline scfw/cli.py:174]: I did not know you can do that.. 💯

File: scfw/cli.py [hunk 2]
```diff
@@ -90,8 +132,16 @@ def _parse_command_line(argv: list[str]) -> tuple[Optional[Namespace], str]:
...
+            case True, False:
+                args_dict = vars(args)
+                args_dict["command"] = argv[hinge:]
+            case False, True:
```
Comment (sobregosodd) [inline scfw/cli.py:179]: Why pass? What's the behaviour here?

File: scfw/cli.py [hunk 3]
```diff
@@ -90,8 +132,16 @@ def _parse_command_line(argv: list[str]) -> tuple[Optional[Namespace], str]:
...
+                pass
+            case _:
+                raise ArgumentError
```
Comment (sobregosodd) [inline scfw/cli.py:183]: Same here, i would add a comment to clarify

File: scfw/configure.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        A `list[inquirer.Question]` of configuration questions to prompt the user
+        for answers to. The list is guaranteed to be nonempty.
+    """
+    has_dd_api_key = os.getenv(DD_API_KEY_VAR) is not None
+
+    questions = [
+        inquirer.Confirm(
+            name="alias_pip",
+            message="Would you like to set a shell alias to run all pip commands through the firewall?",
+            default=False
```
Comment (sobregosodd) [inline scfw/configure.py]: if the user is using config, i would use this as `True`

File: scfw/configure.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        ),
+        inquirer.Confirm(
+            name="alias_npm",
+            message="Would you like to set a shell alias to run all npm commands through the firewall?",
+            default=False
```
Comment (sobregosodd) [inline scfw/configure.py]: same

File: scfw/configure.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+        config += '\nalias npm="scfw run npm"'
+    if answers["dd_api_key"]:
+        config += f'\nexport {DD_API_KEY_VAR}="{answers["dd_api_key"]}"'
+    if answers["dd_log_level"]:
+        config += f'\nexport {DD_LOG_LEVEL_VAR}="{answers["dd_log_level"]}"'
+
+    return enclose(config) if config else config
+
+
+def _update_config_file(config_file: Path, config: str) -> None:
```
Comment (sobregosodd) [inline scfw/configure.py:130]: An easier way would be to:
- read the config into memory (it's a config not a data file)
- if contains `_BLOCK_START` then  re.sub replace from `_BLOCK_START` to `_BLOCK_END` with contents
- if not, append contents
- write to a tmp and rename

File: README.md [hunk 7]
```diff
@@ -50,33 +51,26 @@ The supply-chain firewall is compatible with `pip >= 22.2` and `npm >= 7.0`.  In
...
 
 ## Datadog Logs integration
 
-The supply-chain firewall can optionally send logs of blocked and successful installations to Datadog.
+The supply-chain firewall can optionally send logs of blocked and successful installations to Datadog.  All that is needed is a Datadog API key.
 
 ![scfw datadog log](images/datadog_log.png)
 
-To opt in, set the environment variable `DD_API_KEY` to your Datadog API key, either directly in your shell environment or in a `.env` file in the current working directory.  A logging level may also be selected by setting the environment variable `SCFW_DD_LOG_LEVEL` to one of `ALLOW`, `ABORT` or `BLOCK`.  The `BLOCK` level only logs blocked installations, `ABORT` logs blocked and aborted installations, and `ALLOW` logs these as well as successful installations.  The `BLOCK` level is set by default, i.e., when `SCFW_DD_LOG_LEVEL` is either not set or does not contain a valid log level.
```
Comment (sobregosodd) [inline README.md:79]: I'm more in favor of keep explaining the variables to let the user understand how to do this without having to run the `configure` method. (I prefer avoid interactive commands if possible)

File: scfw/cli.py (hunk 2 above)
Comment (ikretz) [reply scfw/cli.py:179]: The idea is you only want to allow a package manager command when the user has selected the "run" subcommand.  This second case is "(not 'run') AND (no package manager command)", which is allowed and correct but no additional action is needed.

I'll add a comment.

File: scfw/cli.py (hunk 1 above)
Comment (ikretz) [reply scfw/cli.py:174]: Oh yeah, it's awesome.  I wish every language had `match` tbh

File: README.md (hunk 7 above)
Comment (ikretz) [reply README.md:79]: Ahh, good point.  If I have time, one enhancement I would like to add to the `configure` subcommand is to provide all config options as optional arguments (I also hate interactive commands).  But I will add this section back in.

File: scfw/cli.py (hunk 2 above)
Comment (ikretz) [reply scfw/cli.py:179]: Comment added in [00a289d](https://github.com/DataDog/supply-chain-firewall/pull/31/commits/00a289d76aea30880375312970c29716bc43c0af)

File: scfw/cli.py (hunk 3 above)
Comment (ikretz) [reply scfw/cli.py:183]: Comment added in [00a289d](https://github.com/DataDog/supply-chain-firewall/pull/31/commits/00a289d76aea30880375312970c29716bc43c0af)

File: README.md (hunk 7 above)
Comment (ikretz) [reply README.md:79]: Added back in [00a289d](https://github.com/DataDog/supply-chain-firewall/pull/31/commits/00a289d76aea30880375312970c29716bc43c0af)

File: scfw/configure.py (hunk 4 above)
Comment (ikretz) [reply scfw/configure.py]: Fixed in [0231a26](https://github.com/DataDog/supply-chain-firewall/pull/31/commits/0231a2699da901827983ac14f858e55e1efbc442)

File: scfw/configure.py (hunk 5 above)
Comment (ikretz) [reply scfw/configure.py]: Fixed in [0231a26](https://github.com/DataDog/supply-chain-firewall/pull/31/commits/0231a2699da901827983ac14f858e55e1efbc442)

File: scfw/configure.py (hunk 6 above)
Comment (ikretz) [reply scfw/configure.py:130]: Agreed.  Changed in [348f955](https://github.com/DataDog/supply-chain-firewall/pull/31/commits/348f9550ee8b4f75dd7db1cc2bcc0b4a0b379835).

File: scfw/configure.py [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+        config: The new configuration to write.
+    """
+    def enclose(config: str) -> str:
+        return f"{_BLOCK_START}{config}\n{_BLOCK_END}"
+
+    with open(config_file) as f:
+        contents = f.read()
+
+    updated = re.sub(f"{_BLOCK_START}(.*?){_BLOCK_END}", enclose(config), contents, flags=re.DOTALL)
+    if updated == contents:
```
Comment (sobregosodd) [inline scfw/configure.py]: there is a small problem with this approach: if the config file does have the settings but they are the replaced content is the same as before the `re.sub` (the user typed the exact same config settings two times), you will end up appending the config settings again.

Comment (ikretz) [reply scfw/configure.py]: Great catch, thanks!  Fixed in [7dd8192](https://github.com/DataDog/supply-chain-firewall/pull/31/commits/7dd81928518e110e260f7acc046097437e081bd5)

**Pull request merged**